README.md
# birdears

[![Maintenance](https://img.shields.io/maintenance/yes/2017.svg?style=flat-square)](https://github.com/iacchus/birdears/issues/new?title=Is+birdears+still+maintained&body=Please+file+an+issue+if+the+maintained+button+says+no)
[![Travis Build Status](https://img.shields.io/travis/iacchus/birdears.svg?style=flat-square&label=build)](https://travis-ci.org/iacchus/birdears)
[![Coveralls](https://img.shields.io/coveralls/iacchus/birdears.svg?style=flat-square&label=coverage)](https://coveralls.io/github/iacchus/birdears)
[![Awesome Sheet Music](https://img.shields.io/badge/awesome-sheet%20music-blue.svg?style=flat-square&logoWidth=14;&logo=data%3Aimage%2Fpng%3Bbase64%2CiVBORw0KGgoAAAANSUhEUgAAAA4AAAAOCAYAAAAfSC3RAAAABmJLR0QA%2FwD%2FAP%2BgvaeTAAAACXBIWXMAAD2EAAA9hAHVrK90AAAAB3RJTUUH4QYVEQ4dGSq4mgAAAuVJREFUKM8FwUtoHGUAB%2FD%2F983s7CSb7s6%2BsrtJtqbEJA21llgQi4VaCL5QRKXQg6JXEUTqrfQi9CTozYPeUgrtxceh1kbEEEtBeslzTbdrup109r2zMzvZ2Xl8s9%2F4%2B5Hvfg4vOGpzqJf3aCpY%2FfTMi5OvE%2B6mdx%2Fud0YjP5PNxpiuh6X9%2F3I%2F6mzcScvT%2BvjYCUucknhkYSnvNYTvr5169YNLCEMAHKfPvZxwrBZaqoaGWn%2BNBNWlJ4dzn3147n2totckelwZhdr%2B9U%2FOvnnx0kRSgTJdhOBpIPYzSMyE7DaQJEcoFOO5lFK%2BeszcjxblGU%2BUzHosO5%2B6Ek2kQUYuupt3cXxxEoTIYN0WFDGHNnGwvVFGLJ1eXqv%2B8dzJPK%2FRhvbLKwtnzoI7BvyBjqBzCGEiC5ougAgCCJUAymHZPuYWc8J27cGV3c76jOgM9FlwjpE7APMZAscHH%2FYQsBgAhqBvYCyeREgRbm%2FV3aXixLwf6DFxqpCIDpqPQMQIjNoTBLaN9uYO7v30T9h72sHlL1fIV1%2Bvh0mBsp16na6czmUuJM5XaM%2Fo5UEp9JqGeGEO7tBHfGYW73x8kay8%2FTwRvADL02PIClw6shkxbd8sxl%2Fo0yghVl%2FvIyoL0DY30G8ZcKt7kDnDVDGHQbOFNAFZzlC4ozDiub6SofGAWsNIZWh10X7WwtODNsyeh15ZhfVYw6M%2Ft9Do%2B1jdMsPAC%2BEEHCyAy5WCT5VE%2FqDyr4rh0IZlu%2BygZg%2FdaAw2O0KTUWh%2FVXBe4kT1KVgIeAPh1rHxJBdufHsnKJXvXK7slIV61SodNI7iswVRDgnB%2FEuz8IiDvs1xs2yH1Q099oaZfyj5lkRVf4Ta4%2B69vuF6ajeodkRJ2tzroVRq4%2F5v2xiTFahd115ITn5eu23L5on3mBn5O0UNTxB2m%2FIDdZD5hiUW7qcyhd%2B%2F%2BHUNc2%2B9i8OBwFfXDo11Hfjho2t3I4tRIRoYNBAV738fRoHSSCa2GwAAAABJRU5ErkJggg%3D%3D)](https://iacchus.github.io/awesome-sheet-music/)

[![GitHub (pre-)release](https://img.shields.io/github/release/iacchus/birdears/all.svg?style=flat-square)](https://github.com/iacchus/birdears/releases)
[![PyPI Status](https://img.shields.io/pypi/status/birdears.svg?style=flat-square&label=pypi-status)](https://pypi.python.org/pypi/birdears)
[![PyPI Version](https://img.shields.io/pypi/v/birdears.svg?style=flat-square)](https://pypi.python.org/pypi/birdears)
[![PyPI Python Versions](https://img.shields.io/pypi/pyversions/birdears.svg?style=flat-square)](https://pypi.python.org/pypi/birdears)
[![Documentation Status](https://img.shields.io/badge/readthedocs-latest-orange.svg?style=flat-square)](https://birdears.readthedocs.io/en/latest/)

<!-- TOC depthFrom:2 depthTo:6 withLinks:1 updateOnSave:1 orderedList:0 -->

- [Introduction](#introduction)
	- [birdears](#birdears)
	- [What is musical ear training](#what-is-musical-ear-training)
	- [Features](#features)
- [Installing](#installing)
	- [Installing the dependencies](#installing-the-dependencies)
		- [Arch Linux](#arch-linux)
	- [Installing birdears](#installing-birdears)
		- [In-depth installation](#in-depth-installation)
- [Running](#running)
- [Keybindings](#keybindings)
- [Documentation](#documentation)
- [Contributing](#contributing)

<!-- /TOC -->

For the support chat, please `/join` [`#birdears`](https://webchat.freenode.net/?randomnick=1&channels=%23birdears&uio=MTY9dHJ1ZSYxMT0yNDY57) channel on freenode (`chat.freenode.net/6697` - ssl).

![birdears screencast](use.gif)

## Introduction


### birdears

`birdears` is a software written in Python 3 for ear training for musicians
(musical intelligence, transcribing music, composing). It is a clone of the
method used by [Funcitional Ear Trainer](https://play.google.com/store/apps/details?id=com.kaizen9.fet.android) app for Android.

It comes with four modes, or four kind of exercises, which are: `melodic`, `harmonic`, `dictation` and `instrumental`.

In resume, with the *melodic* mode two notes are played one after the other and you have to guess the interval; with the
`harmonic` mode, two notes are played simoutaneously (harmonically) and you should guess the interval.

With the *dictation* mode, more than 2 notes are played (*ie*., a melodic dictation) and you should tell what are the intervals
between them.

With the *instrumental* mode, it is a like the *dictation*, but you will be expected to play the notes on your instrument, *ie*.,
birdears will not wait for a typed reply and you should prectice with your own judgement. The melody can be repeat any times and
you can have as much time as you want to try it out.

### What is musical ear training

*this needs to be written. The method.*

### Features

* questions
* pretty much configurable
* load from config file
* you can make your own presets
* can be used interactively *(docs needed)*
* can be used as a library *(docs needed)*

## Installing

### Installing the dependencies

`birdears` depends only on `python >= 3.5` and `sox` (which should be installed by your distribution's package manager,
supposing you're using linux, and which provides the `play` command.) 

#### Arch Linux

```
sudo pacman -Syu sox python python-pip
```

### Installing birdears

`pip install --user --upgrade --no-cache-dir birdears`

#### In-depth installation

You can choose to use a virtualenv to use birdears; this should give you an idea on how to setup one virtualenv.

You should first install virtualenv (for python3) using your distribution's package (supposing you're on linux),
then on terminal:

```
virtualenv -p python3 ~/.venv # use the directory ~/.venv/ for the virtualenv

source ~/.venv/bin/activate   # activate the virtualenv; this should be done
                              # every time you may want to run the software
                              # installed here.

pip3 install birdears         # this will install the software

birdears --help               # and this will run it
```

## Running

After installing just run:

`birdears --help`

## Keybindings

The following keyboard diagrams should give you an idea on how the keybindings work. Please
note how the keys on the line from `z` (*unison*) to `,` (comma, *octave*) represent the notes
that are *natural* to the mode, and the line above represent the chromatics.

Also, for exercises with two octaves, the **uppercased keys represent the second octave**. For
example, `z` is *unison*, `,` is the *octave*, `Z` (uppercased) is the *double octave*. The same
for all the other intervals.

### Ionian (Major)

These are the keybindings for the **Ionian (Major) Scale**; black keys are the chromatic notes.

![birdears - ionian(major) keybindings](docs/keybindings/ionian.png)

### Dorian

![birdears - dorian keybindings](docs/keybindings/dorian.png)

### Phrygian

![birdears - phryigian keybindings](docs/keybindings/phrygian.png)

### Lydian

![birdears - lydian keybindings](docs/keybindings/lydian.png)

### Mixolydian

![birdears - mixolydian keybindings](docs/keybindings/mixolydian.png)

### Aeolian (minor)

![birdears - aeolian keybindings](docs/keybindings/minor.png)

### Locrian

![birdears - locrian(minor) keybindings](docs/keybindings/locrian.png)

## Advanced

![birdears - advanced keybindings](docs/keybindings/full-advanced.png)

*this is still being improved*

Legend for the keys on the diagram above:

| Text  Format       | Scale Direction | Octave                             |
|--------------------|-----------------|------------------------------------|
| blue (bold italic) | descending      | second octave (shift or caps lock) |
| pink (bold)        | descending      | first octave                       |
| black/white (bold) | ascending       | first octave                       |
| black (italic)     | ascending       | second octave (shift or caps lock) |

**White keys** are the diatonic notes, **black keys** are the chromatic ones.

Descendent mode are usable for exercises with `-d` or `--descendent`.

Chromatic keys are usable for exercises with `-c` or `--chromatic`.

Second octave is usable for exercises with `-n 2` or `--n_octaves 2`

## Documentation

Full documentation for this software is available at birdears [Read The Docs](https://birdears.readthedocs.io/en/latest/)
and also in [PDF format](https://github.com/iacchus/birdears/raw/master/docs/sphinx/_build/latex/birdears.pdf).

## Contributing

Those who want to contribute to this project can read [CONTRIBUTING.md](CONTRIBUTING.md).

## Etc

the screencast was recorded with a command similar to

```
COLUMNS=120 LINES=36 ttyrec
seq2gif -s 2 -i ttyrecord -w 120 -h 36 -o use.gif
```

[ttyrec](https://aur.archlinux.org/packages/ttyrec/) and [seq2gif](https://github.com/saitoha/seq2gif)

keyboard layouts were generated with http://www.keyboard-layout-editor.com/


CHANGELOG
0.1.5

* Fixed bug with prequestion method 'tonic_only' with dictation
* Changed sox play command to fix bug with short notes (experimental)



birdears/__init__.py
"""
birdears provides facilities to building musical ear training exercises.
"""

from .logger import logger

DEBUG = False

CHROMATIC_SHARP = ('C', 'C#', 'D', 'D#', 'E', 'F', 'F#', 'G', 'G#', 'A', 'A#',
                   'B')
"""tuple: Chromatic notes names using sharps.

A mapping of the chromatic note namesu sing sharps"""

CHROMATIC_FLAT = ('C', 'Db', 'D', 'Eb', 'E', 'F', 'Gb', 'G', 'Ab', 'A', 'Bb',
                  'B')
"""tuple: Chromatic notes names using flats.

A mapping of the chromatic note names using flats."""

KEYS = ('C', 'C#', 'Db', 'D', 'D#', 'Eb', 'E', 'F', 'F#', 'Gb', 'G', 'G#',
        'Ab', 'A', 'A#', 'Bb', 'B')
"""tuple: Allowed keys

These are the allowed keys for exercise as comprehended by birdears."""

CIRCLE_OF_FIFTHS = [
        ('C', 'G', 'D', 'A', 'E', 'B', 'Gb', 'Db', 'Ab', 'Eb', 'Bb', 'F'),
        ('C', 'F', 'Bb', 'Eb', 'Ab', 'C#', 'F#', 'B', 'E', 'A', 'D', 'G')
    ]
"""list of tuples: Circle of fifths.

These are the circle of fifth in both directions."""

INTERVALS = (
    (0, 'P1', 'Perfect Unison'),
    (1, 'm2', 'Minor Second'),
    (2, 'M2', 'Major Second'),
    (3, 'm3', 'Minor Third'),
    (4, 'M3', 'Major Third'),
    (5, 'P4', 'Perfect Fourth'),
    (6, 'A4', 'Augmented Fourth'),
    (7, 'P5', 'Perfect Fifth'),
    (8, 'm6', 'Minor Sixth'),
    (9, 'M6', 'Major Sixth'),
    (10, 'm7', 'Minor Seventh'),
    (11, 'M7', 'Major Seventh'),
    (12, 'P8', 'Perfect Octave'),
    (13, 'A8', 'Minor Ninth'),
    (14, 'M9', 'Major Ninth'),
    (15, 'm10', 'Minor Tenth'),
    (16, 'M10', 'Major Tenth'),
    (17, 'P11', 'Perfect Eleventh'),
    (18, 'A11', 'Augmented Eleventh'),
    (19, 'P12', 'Perfect Twelfth'),
    (20, 'm13', 'Minor Thirteenth'),
    (21, 'M13', 'Major Thirteenth'),
    (22, 'm14', 'Minor Fourteenth'),
    (23, 'M14', 'Major Fourteenth'),
    (24, 'P15', 'Perfect Double-octave'),
    (25, 'A15', 'Minor Sixteenth'),
    (26, 'M16', 'Major Sixteenth'),
    (27, 'm17', 'Minor Seventeenth'),
    (28, 'M17', 'Major Seventeenth'),
    (29, 'P18', 'Perfect Eighteenth'),
    (30, 'A18', 'Augmented Eighteenth'),
    (31, 'P19', 'Perfect Nineteenth'),
    (32, 'm20', 'Minor Twentieth'),
    (33, 'M20', 'Major Twentieth'),
    (34, 'm21', 'Minor Twenty-first'),
    (35, 'M21', 'Major Twenty-first'),
    (36, 'P22', 'Perfect Triple-octave')
)
"""tuple of tuples: Data representing intervals.

A tuple of tuples representing data for the intervals with format
(semitones, short name, full name)."""

CHROMATIC_TYPE = (0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12)
"""tuple: A map of the chromatic chromatic scale.

A map of the the semitones which compound the chromatic scale."""

DIATONIC_MODES = {
    'major': (0, 2, 4, 5, 7, 9, 11, 12),
    'dorian': (0, 2, 3, 5, 7, 9, 10, 12),
    'phrygian': (0, 1, 3, 5, 7, 8, 10, 12),
    'lydian': (0, 2, 4, 6, 7, 9, 11, 12),
    'mixolydian': (0, 2, 4, 5, 7, 9, 10, 12),
    'minor': (0, 2, 3, 5, 7, 8, 10, 12),
    'locrian': (0, 1, 3, 5, 6, 8, 10, 12),
}
"""dict of tuples: A map of the diatonic scale.

A mapping of the semitones which compound each of the greek modes.
"""

INTERVAL_INDEX = {
    1: [0],
    2: [1, 2],
    3: [3, 4],
    4: [5, 6],
    5: [6, 7],
    6: [8, 9],
    7: [10, 11],
    8: [12]
}
"""dict of lists: A mapping of semitones of each interval.

A mapping of semitones which index to each interval name, major/minor,
perfect, augmented/diminished"""

# MAJOR keyboard keys (with chromatics)
# keyindex for major and chromatic major context
#  s d   g h j        IIb  IIIb       Vb VIb  VIIb
# z x c v b n m  <-  I   II   III  IV   V   VI   VII
#
#  -- SHIFT meaning an octave higher

# MINOR keyboard keys (with chromatics)
# keyindex for minor and chromatic minor context
#
#  s   f g   j k   eg.:      a#   c# d#    f# g#
# z x c v b n m    -------  a  b c  d  e  f  g


# it is better if this is made progamatically
# note that all diatonic are in keys zxcvvbnm, and uppercase
KEYBOARD_INDICES = {
    'diatonic': {
        'ascending': {
            'minor': 'z xc v bn m Z XC V BN M,',
            'major': 'z x cv b n mZ X CV B NM,'
        },
        # FIXME
        'descending': {
            'minor': 'z xc v bn m Z XC V BN M,',
            'major': 'z x cv b n mZ X CV B NM,'
        },
    },
    'chromatic': {
        'ascending': {
            'major':      "zsxdcvgbhnjm,SXDCVGBHNJMZ",
            'dorian':     "zsxcfvgbhnmk,SXCFVGBHNMKZ",
            'phrygian':   "zxdcfvgbnjmk,XDCFVGBNJMKZ",
            'lydian':     "zsxdcfvbhnjm,SXDCFVBHNJMZ",
            'mixolydian': "zsxdcvgbhnmk,SXDCVGBHNMNZ",
            'minor':      "zsxcfvgbnjmk,SXCFVGBNJMKZ",
            'locrian':    "zxdcfvbhnjmk,XDCFVBHNJMKZ",
        },
        'descending': {
            'major':      ",mjnhbgvcdxszMJNHBGVCDXSZ",
            'dorian':     ",kmnhbgvfcxszKMNHBGVFCXSZ",
            'dorian':     ",kmnhbgvfcxszKMNHBGVFCXSZ",
            'phrygian':   ",kmjnbgvfcdxzKMJNBGVFCDXZ",
            'lydian':     ",mjnhbvfcdxszMJNHBVFCDXSZ",
            'mixolydian': ",kmnhbgvcdxszKMNHBGVCDXSZ",
            'minor':      ",kmjnbgvfcxszKMJNBGVFCXSZ",
            'locrian':    ",kmjnhbvfcdxzKMJNHBVFCDXZ",
        }
    }
}

# how many steps to resolve on tonic,
# after which resolves on tonic octave
# (begins on 0 as it is an index)
# 5 is tritone; then
# [0 1 2 3 4 5] resolves below (tonic)
# and [6 7 8 9 10 11] resolves above (octave)

MAX_SEMITONES_RESOLVE_BELOW = 5


# http://code.activestate.com/recipes/134892/
class _Getch:
    """Gets a single character from standard input.  Does not echo to the
screen."""

    def __init__(self):
        try:
            self.impl = _GetchWindows()
        except ImportError:
            self.impl = _GetchUnix()

    def __call__(self): return self.impl()


class _GetchUnix:
    def __init__(self):
        import tty
        import sys

    def __call__(self):
        import sys
        import tty
        import termios
        fd = sys.stdin.fileno()
        old_settings = termios.tcgetattr(fd)
        try:
            tty.setraw(sys.stdin.fileno())
            ch = sys.stdin.read(1)
        finally:
            termios.tcsetattr(fd, termios.TCSADRAIN, old_settings)
        return ch


class _GetchWindows:
    def __init__(self):
        import msvcrt

    def __call__(self):
        import msvcrt
        return msvcrt.getch()
